src/t_zset.c

```diff
@@ -1165,7 +1166,43 @@ unsigned long zsetLength(const robj *zobj) {
     return length;
 }
 
+/* Factory method to return a zset.
+ *
+ * The size hint indicates approximately how many items will be added,
+ * and the value len hint indicates the approximate individual size of the added elements,
+ * they are used to determine the initial representation. */
+robj *zsetTypeCreate(size_t size_hint, size_t val_len_hint) {
+    if (size_hint <= server.zset_max_listpack_entries &&
+        val_len_hint <= server.zset_max_listpack_value)
+    {
+        return createZsetListpackObject();
+    }
+
+    robj *zobj = createZsetObject();
+    zset *zs = zobj->ptr;
+    dictExpand(zs->dict, size_hint);
```

Comment:
i see that dictExpand can also shrink, and unlike dictResize it doesn't protect against going below DICT_HT_INITIAL_SIZE.
am i missing anything? if not, maybe we need to revise many calls, or modify dictExpand.

another thing i noticed is that dictExpand leaves the dict in a rehashing state, even if it was previously empty.
it's not that bad since all the new records will be added to the new HT, and it'll quickly finish rehashing and dismiss the old HT.
maybe we wanna fix it (improve dictExpand when the dict is empty), or maybe we wanna add a test (using DEBUG HTSTATS-KEY) to validate that after the insertion we're not in a rehashing state.

WDYT?

Reply:
yean, i think there are good points, i can get it done. Let me summarize the changes:
1. modify dictExpand to check DICT_HT_INITIAL_SIZE. i can move the DICT_HT_INITIAL_SIZE check from dictResize to dictExpand
2. improve dictExpand when the dict is empty. in _dictExpand, if we see `d->ht_used[0] == 0`, we just copy the new ht onto the old one. (or we can just call dictRehash(d,1))
3. add a test to validate that after the insertion we're not in a rehashing state.

Does this change need to be in a new PR (for better review or backport) or should I just add it in this PR?

Reply:
good point about the backport. if we do wanna consider part of this an important fix, i wouldn't want to backport it with any optimizations, or tests that rely on recent changes.

on the other hand, if we can confirm that it's not a major issue, we can just leave this for 7.2.
so for that i'm interested in the behavior before any fix.
1. if we dictExpand an empty dict, and then insert a few records, will it already be out of resharhing state when the insertions are over? i imagine it'll be done by the second insertion. (which will mean that this fix is just a minor inefficiency fix)
2. is there any real damage of doing dictExpand to a lower value (e.g. 1), or just an inefficiency that will cause it to be re-hashed soon when adding a few more fields?

can you make these checks for me?

Reply:
regarding your commit https://github.com/redis/redis/commit/781353d0d78ea9ec3522c0e8749998d91da04107
i was thinking that it's nicer to re-use this (add another entry condition, and a line to release the old allocation)

```c
    /* Is this the first initialization? If so it's not really a rehashing
     * we just set the first hash table so that it can accept keys. */
    if (d->ht_table[0] == NULL) {
```

instead of adding this at the end.
```c
    /* If the dict was previously empty, then we can end the rehash. */
    if (d->ht_used[0] == 0) _dictSwapAfterRehash(d);
```

Reply:
> if we dictExpand an empty dict, and then insert a few records, will it already be out of resharhing state when the insertions are over? i imagine it'll be done by the second insertion. (which will mean that this fix is just a minor inefficiency fix)

because the dictionary was just created (empty dict), ht_table[0] == NULL, so it will go here:
```
    if (d->ht_table[0] == NULL) {
        d->ht_size_exp[0] = new_ht_size_exp;
        d->ht_used[0] = new_ht_used;
        d->ht_table[0] = new_ht_table;
        return DICT_OK;
    }
```

> is there any real damage of doing dictExpand to a lower value (e.g. 1), or just an inefficiency that will cause it to be re-hashed soon when adding a few more fields?

there is no harm. if we dictExpand(1), because `_dictNextExp(1)` function, we will at least set the size to 4.

so we can drop the commit...

Reply:
> there is no harm. if we dictExpand(1), because _dictNextExp(1) function, we will at least set the size to 4.

ohh, i was hoping that i'm missing something.. great!.

> because the dictionary was just created (empty dict), ht_table[0] == NULL, so it will go here:

ohh, so dictCreate doesn't allocate any table, and in fact it is implicitly allocated on the first call to dictAdd (by _dictExpandIfNeeded)?
so a call to dictExpand before adding any elements would just do the right thing, and no change is needed.

well, sorry for bothering you with my paranoia without checking the facts...

Reply:
@enjoy-binbin please ack that the PR is ready for merge.

Reply:
> ohh, so dictCreate doesn't allocate any table, and in fact it is implicitly allocated on the first call to dictAdd (by _dictExpandIfNeeded)?
so a call to dictExpand before adding any elements would just do the right thing, and no change is needed.
well, sorry for bothering you with my paranoia without checking the facts..

yes, and yes, no worry. i think this PR is ready for merge